### PR TITLE
nodejs: fix build with clang 16

### PR DESCRIPTION
**Pull request by reckenrode**
## Description of changes

Node.js requires severals fixes to build with clang 16:

* Use `posix_memalign` on x86_64-darwin. Otherwise, libc++ will try to use `aligned_alloc`, which is not available until 10.15. This is done by defining it as using the 10.13 SDK even though the 10.12 SDK is sufficient;
* Use clang <16 with v14 and v16. The v18 patch doesn’t apply, but `-Wenum-constexpr-conversion` may become a hard error in a future clang release. Using an older version ensures it builds until these versions are dropped from nixpkgs.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/web/nodejs/v16.nix [hunk 1]
```diff
@@ -1,8 +1,20 @@
-{ callPackage, openssl, python3, fetchpatch, enableNpm ? true }:
+{ callPackage, lib, overrideCC, pkgs, buildPackages, openssl, python3, fetchpatch, enableNpm ? true }:
 
 let
+  # Clang 16+ cannot build Node v14 due to -Wenum-constexpr-conversion errors.
+  # Use an older version of clang with the current libc++ for compatibility (e.g., with icu).
+  ensureCompatibleCC = packages:
```
Comment (toonn) [inline pkgs/development/web/nodejs/v16.nix:6]: Does the node infra have something like a `common.nix` where this would fit?

**Review by toonn — COMMENTED** [review]
Looks pretty good to me, only one nit. Still trying to run test builds.

File: pkgs/development/web/nodejs/v16.nix (hunk 1 above)
Comment (reckenrode) [reply pkgs/development/web/nodejs/v16.nix:6]: It does, but I don’t want to put stuff in there that’s only needed until the deprecated versions are removed.

**Review by toonn — COMMENTED** [review]
Rebasing this on the LLVM 16 bump I'm getting a permission error about `env` not being permitted as an interpreter?

Last lines of the build log:
```
  LD_LIBRARY_PATH=/private/tmp/nix-build-nodejs-18.17.1.drv-0/node-v18.17.1/out/
Release/lib.host:/private/tmp/nix-build-nodejs-18.17.1.drv-0/node-v18.17.1/out/R
elease/lib.target:$LD_LIBRARY_PATH; export LD_LIBRARY_PATH; cd ../.; mkdir -p /p
rivate/tmp/nix-build-nodejs-18.17.1.drv-0/node-v18.17.1/out/Release/obj/gen/src;
 python tools/inspector_protocol/convert_protocol_to_json.py src/inspector/node_
protocol.pdl "/private/tmp/nix-build-nodejs-18.17.1.drv-0/node-v18.17.1/out/Rele
ase/obj/gen/src/node_protocol.json"                                             
  rm -f /private/tmp/nix-build-nodejs-18.17.1.drv-0/node-v18.17.1/out/Release/li
bbase64.a && ./gyp-mac-tool filter-libtool libtool  -static -o /private/tmp/nix-
build-nodejs-18.17.1.drv-0/node-v18.17.1/out/Release/libbase64.a /private/tmp/ni
x-build-nodejs-18.17.1.drv-0/node-v18.17.1/out/Release/obj.target/base64/deps/ba
se64/base64/lib/arch/generic/codec.o /private/tmp/nix-build-nodejs-18.17.1.drv-0
/node-v18.17.1/out/Release/obj.target/base64/deps/base64/base64/lib/tables/table
s.o /private/tmp/nix-build-nodejs-18.17.1.drv-0/node-v18.17.1/out/Release/obj.ta
rget/base64/deps/base64/base64/lib/codec_choose.o /private/tmp/nix-build-nodejs-
18.17.1.drv-0/node-v18.17.1/out/Release/obj.target/base64/deps/base64/base64/lib/lib.o /private/tmp/nix-build-nodejs-18.17.1.drv-0/node-v18.17.1/out/Release/obj.target/base64/deps/base64/base64/lib/arch/neon32/codec.o /private/tmp/nix-build-nodejs-18.17.1.drv-0/node-v18.17.1/out/Release/obj.target/base64/deps/base64/base64/lib/arch/neon64/codec.o
In file included from <built-in>:440:
<command line>:32:9: warning: '__ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__' macro redefined [-Wmacro-redefined]
#define __ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ 101300
        ^
<built-in>:431:9: note: previous definition is here
#define __ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ 101500
        ^
/nix/store/0iqd25xfcld27z9kg92qdayj872rw2k9-bash-5.2-p15/bin/bash: ./gyp-mac-tool: /usr/bin/env: bad interpreter: Operation not permitted
make[1]: *** [deps/base64/base64.target.mk:174: /private/tmp/nix-build-nodejs-18.17.1.drv-0/node-v18.17.1/out/Release/libbase64.a] Error 126
make[1]: *** Waiting for unfinished jobs....
21 warnings generated.
1 warning generated. 
23 warnings generated.
rm bd7fbf3b03451c23263e00021ee01d702a3b9540.intermediate c5009241ae6960ccbf1b3255e5c7cbad052383eb.intermediate
make: *** [Makefile:134: node] Error 2
error: builder for '/nix/store/ag1bvf2wxfbwz0559dihhzhavd1blcmg-nodejs-18.17.1.drv' failed with exit code 2
```
Might still be doing something wrong, I don't remember this error from when I looked at this earlier but I can't seem to get around it now.

EDIT: The relation to `env` actually seems tenuous. Sometimes it's the bad interpreter error, other times it's just make error 2.

**Comment by reckenrode** [comment]
I rebased on staging and dropped the node v18 commit. Upstream has cherry-picked the required commits to fix building with clang 16.

**Review by toonn — APPROVED** [review]
LGTM

**Pull request merged**